### PR TITLE
Rewrite it in Rust

### PR DESCRIPTION
**Pull request by teohhanhui**
As discussed in https://github.com/slp/krun/pull/1#issuecomment-2076839288

---

Bugs fixed:
- [x] Current implementation of `krun` incorrectly passes own arguments to `krun-guest`, e.g.
    ```
    teohhanhui@Han-MacBook-Air:~/Projects/slp/krun$ ./krun --net=PASST box64 --help
    No IPv6 nameserver available for NDP/DHCPv6
    Couldn't execute '--net=PASST' inside the vm: No such file or directory
    ```
- [x] Stop using hardcoded `/tmp/passt_1.socket`
- [x] When forking `krun-guest` to run `dhclient` in child process, stdin/stdout/stderr were illegally closed.
    > Usage of `close()` on file descriptors STDIN_FILENO, STDOUT_FILENO, or
    > STDERR_FILENO should immediately be followed by an operation to reopen these
    > file descriptors. Unexpected behavior will result if any of these file
    > descriptors is left in a closed state
    
    See https://pubs.opengroup.org/onlinepubs/9699919799/functions/close.html#tag_16_67_07
- [x] `File exists (os error 17)` when trying to create the temp directory for `XDG_RUNTIME_DIR` after the first run.
   Use properly unique temp dir names for each run.
---

Dependency tree with `cargo tree -e normal --no-dedupe` (excluding `build-dependencies`):
```
krun v0.1.0 (/home/teohhanhui/Projects/slp/krun/crates/krun)
├── anyhow v1.0.82
├── bpaf v0.9.11
├── libkrun v1.8.1 (/home/teohhanhui/Projects/slp/krun/crates/libkrun)
├── nix v0.28.0
│   ├── bitflags v2.5.0
│   ├── cfg-if v1.0.0
│   └── libc v0.2.153
├── rustix v0.38.34
│   ├── bitflags v2.5.0
│   └── linux-raw-sys v0.4.13
├── tracing v0.1.40
│   ├── pin-project-lite v0.2.14
│   ├── tracing-attributes v0.1.27 (proc-macro)
│   │   ├── proc-macro2 v1.0.81
│   │   │   └── unicode-ident v1.0.12
│   │   ├── quote v1.0.36
│   │   │   └── proc-macro2 v1.0.81
│   │   │       └── unicode-ident v1.0.12
│   │   └── syn v2.0.60
│   │       ├── proc-macro2 v1.0.81
│   │       │   └── unicode-ident v1.0.12
│   │       ├── quote v1.0.36
│   │       │   └── proc-macro2 v1.0.81
│   │       │       └── unicode-ident v1.0.12
│   │       └── unicode-ident v1.0.12
│   └── tracing-core v0.1.32
│       └── once_cell v1.19.0
├── tracing-subscriber v0.3.18
│   ├── matchers v0.1.0
│   │   └── regex-automata v0.1.10
│   │       └── regex-syntax v0.6.29
│   ├── nu-ansi-term v0.46.0
│   │   └── overload v0.1.1
│   ├── once_cell v1.19.0
│   ├── regex v1.10.4
│   │   ├── regex-automata v0.4.6
│   │   │   └── regex-syntax v0.8.3
│   │   └── regex-syntax v0.8.3
│   ├── sharded-slab v0.1.7
│   │   └── lazy_static v1.4.0
│   ├── smallvec v1.13.2
│   ├── thread_local v1.1.8
│   │   ├── cfg-if v1.0.0
│   │   └── once_cell v1.19.0
│   ├── tracing v0.1.40
│   │   ├── pin-project-lite v0.2.14
│   │   ├── tracing-attributes v0.1.27 (proc-macro)
│   │   │   ├── proc-macro2 v1.0.81
│   │   │   │   └── unicode-ident v1.0.12
│   │   │   ├── quote v1.0.36
│   │   │   │   └── proc-macro2 v1.0.81
│   │   │   │       └── unicode-ident v1.0.12
│   │   │   └── syn v2.0.60
│   │   │       ├── proc-macro2 v1.0.81
│   │   │       │   └── unicode-ident v1.0.12
│   │   │       ├── quote v1.0.36
│   │   │       │   └── proc-macro2 v1.0.81
│   │   │       │       └── unicode-ident v1.0.12
│   │   │       └── unicode-ident v1.0.12
│   │   └── tracing-core v0.1.32
│   │       └── once_cell v1.19.0
│   ├── tracing-core v0.1.32
│   │   └── once_cell v1.19.0
│   └── tracing-log v0.2.0
│       ├── log v0.4.21
│       ├── once_cell v1.19.0
│       └── tracing-core v0.1.32
│           └── once_cell v1.19.0
└── utils v0.0.0 (/home/teohhanhui/Projects/slp/krun/crates/utils)
    └── tracing v0.1.40
        ├── pin-project-lite v0.2.14
        ├── tracing-attributes v0.1.27 (proc-macro)
        │   ├── proc-macro2 v1.0.81
        │   │   └── unicode-ident v1.0.12
        │   ├── quote v1.0.36
        │   │   └── proc-macro2 v1.0.81
        │   │       └── unicode-ident v1.0.12
        │   └── syn v2.0.60
        │       ├── proc-macro2 v1.0.81
        │       │   └── unicode-ident v1.0.12
        │       ├── quote v1.0.36
        │       │   └── proc-macro2 v1.0.81
        │       │       └── unicode-ident v1.0.12
        │       └── unicode-ident v1.0.12
        └── tracing-core v0.1.32
            └── once_cell v1.19.0

krun-guest v0.1.0 (/home/teohhanhui/Projects/slp/krun/crates/krun-guest)
├── anyhow v1.0.82
├── bpaf v0.9.11
├── nix v0.28.0
│   ├── bitflags v2.5.0
│   ├── cfg-if v1.0.0
│   └── libc v0.2.153
├── rustix v0.38.34
│   ├── bitflags v2.5.0
│   └── linux-raw-sys v0.4.13
├── tempfile v3.10.1
│   ├── cfg-if v1.0.0
│   ├── fastrand v2.1.0
│   └── rustix v0.38.34
│       ├── bitflags v2.5.0
│       └── linux-raw-sys v0.4.13
├── tracing v0.1.40
│   ├── pin-project-lite v0.2.14
│   ├── tracing-attributes v0.1.27 (proc-macro)
│   │   ├── proc-macro2 v1.0.81
│   │   │   └── unicode-ident v1.0.12
│   │   ├── quote v1.0.36
│   │   │   └── proc-macro2 v1.0.81
│   │   │       └── unicode-ident v1.0.12
│   │   └── syn v2.0.60
│   │       ├── proc-macro2 v1.0.81
│   │       │   └── unicode-ident v1.0.12
│   │       ├── quote v1.0.36
│   │       │   └── proc-macro2 v1.0.81
│   │       │       └── unicode-ident v1.0.12
│   │       └── unicode-ident v1.0.12
│   └── tracing-core v0.1.32
│       └── once_cell v1.19.0
├── tracing-subscriber v0.3.18
│   ├── matchers v0.1.0
│   │   └── regex-automata v0.1.10
│   │       └── regex-syntax v0.6.29
│   ├── nu-ansi-term v0.46.0
│   │   └── overload v0.1.1
│   ├── once_cell v1.19.0
│   ├── regex v1.10.4
│   │   ├── regex-automata v0.4.6
│   │   │   └── regex-syntax v0.8.3
│   │   └── regex-syntax v0.8.3
│   ├── sharded-slab v0.1.7
│   │   └── lazy_static v1.4.0
│   ├── smallvec v1.13.2
│   ├── thread_local v1.1.8
│   │   ├── cfg-if v1.0.0
│   │   └── once_cell v1.19.0
│   ├── tracing v0.1.40
│   │   ├── pin-project-lite v0.2.14
│   │   ├── tracing-attributes v0.1.27 (proc-macro)
│   │   │   ├── proc-macro2 v1.0.81
│   │   │   │   └── unicode-ident v1.0.12
│   │   │   ├── quote v1.0.36
│   │   │   │   └── proc-macro2 v1.0.81
│   │   │   │       └── unicode-ident v1.0.12
│   │   │   └── syn v2.0.60
│   │   │       ├── proc-macro2 v1.0.81
│   │   │       │   └── unicode-ident v1.0.12
│   │   │       ├── quote v1.0.36
│   │   │       │   └── proc-macro2 v1.0.81
│   │   │       │       └── unicode-ident v1.0.12
│   │   │       └── unicode-ident v1.0.12
│   │   └── tracing-core v0.1.32
│   │       └── once_cell v1.19.0
│   ├── tracing-core v0.1.32
│   │   └── once_cell v1.19.0
│   └── tracing-log v0.2.0
│       ├── log v0.4.21
│       ├── once_cell v1.19.0
│       └── tracing-core v0.1.32
│           └── once_cell v1.19.0
└── utils v0.0.0 (/home/teohhanhui/Projects/slp/krun/crates/utils)
    └── tracing v0.1.40
        ├── pin-project-lite v0.2.14
        ├── tracing-attributes v0.1.27 (proc-macro)
        │   ├── proc-macro2 v1.0.81
        │   │   └── unicode-ident v1.0.12
        │   ├── quote v1.0.36
        │   │   └── proc-macro2 v1.0.81
        │   │       └── unicode-ident v1.0.12
        │   └── syn v2.0.60
        │       ├── proc-macro2 v1.0.81
        │       │   └── unicode-ident v1.0.12
        │       ├── quote v1.0.36
        │       │   └── proc-macro2 v1.0.81
        │       │       └── unicode-ident v1.0.12
        │       └── unicode-ident v1.0.12
        └── tracing-core v0.1.32
            └── once_cell v1.19.0

libkrun v1.8.1 (/home/teohhanhui/Projects/slp/krun/crates/libkrun)

utils v0.0.0 (/home/teohhanhui/Projects/slp/krun/crates/utils)
└── tracing v0.1.40
    ├── pin-project-lite v0.2.14
    ├── tracing-attributes v0.1.27 (proc-macro)
    │   ├── proc-macro2 v1.0.81
    │   │   └── unicode-ident v1.0.12
    │   ├── quote v1.0.36
    │   │   └── proc-macro2 v1.0.81
    │   │       └── unicode-ident v1.0.12
    │   └── syn v2.0.60
    │       ├── proc-macro2 v1.0.81
    │       │   └── unicode-ident v1.0.12
    │       ├── quote v1.0.36
    │       │   └── proc-macro2 v1.0.81
    │       │       └── unicode-ident v1.0.12
    │       └── unicode-ident v1.0.12
    └── tracing-core v0.1.32
        └── once_cell v1.19.0
```

Requires https://github.com/containers/libkrun/pull/181

### CODE REVIEW

**Comment by teohhanhui** [comment]
I think we can remove the `errno` dependency by not calling `set_errno` (we're not a library so it's pointless for us to set `errno` anyway...)

If I understand correctly, `errno` was only being assigned to for the side effect in `perror`, i.e. to print out the human-readable error message corresponding to the `errno`.

`rustix` already has us covered for that use case.

**Comment by slp** [comment]
I think it's looking good so far, thanks for working on this. Do you plan to rewrite `krun-guest` too or do you plan on keeping it as C?

**Comment by teohhanhui** [comment]
I plan to rewrite it too, but I'll only find time in the next few days.

**Comment by teohhanhui** [comment]
`krun-guest` is now tested to be working. Only `exec_sommelier` is not implemented yet...

File: src/net.rs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+pub fn connect_to_passt<P>(passt_socket_path: P) -> Result<UnixStream>
+where
+    P: AsRef<Path>,
+{
+    Ok(UnixStream::connect(passt_socket_path)?)
+}
+
+/// Starts `passt`.
+///
+/// This does [fork(2)] and [exec(3)].
```
Comment (teohhanhui) [inline src/net.rs:35]: This is the only manual `fork` + `exec` left...

I think we could probably get rid of it. Just need to look into this.

Comment (slp) [reply src/net.rs:35]: Yes, there's nothing special about running passt as a daemon. It should be possible replacing `fork+exec` with `std::process`.

**Comment by slp** [comment]
@teohhanhui This is looking really good so far! Is there anything else remaining to be done apart for the fork+exec replacement for passt?

**Comment by slp** [comment]
@teohhanhui I'm also thinking about renaming this repo as `krun.legacy` and creating a clean one with just `README.md` and `LICENSE`. Would you mind rebasing your PR to that one?

**Comment by teohhanhui** [comment]
I think there's value in keeping the commit history, but it's your call.

**Comment by teohhanhui** [comment]
> Is there anything else remaining to be done apart for the fork+exec replacement for passt?

There's the hardcoded `/home/{username}` for `HOME` env var. Should we pass in the actual path from `getpwnam_r` to `krun-guest` instead (already used in `krun`)?

There's also the TODO that you've noted about only overriding the mesa driver selection on Asahi Linux. Is it possible for us to set that in `krun` instead? (through `envp` in `krun_set_exec`)

File: Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+[package]
+name = "krun"
+version = "0.1.0"
+authors = ["Sergio Lopez <slp@redhat.com>"]
```
Comment (slp) [inline Cargo.toml:4]: Please add your name here.

File: Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+rust-version = "1.77"
+description = "Run programs from your system in a microVM"
+repository = "https://github.com/slp/krun"
+license = "MIT"
+default-run = "krun"
+
+[dependencies]
+anyhow = "1.0.82"
+arrayvec = "0.7.4"
+bpaf = "0.9.11"
```
Comment (slp) [inline Cargo.toml:15]: Doesn't look like `bpaf` is packaged in Fedora. Unless there's a strong reason to use it, I think switching to something like `clap` would make things easier for us.

Comment (teohhanhui) [reply Cargo.toml:15]: I don't mind packaging `bpaf`. I have a strong dislike for `clap` (it sucks).

File: src/bin/krun.rs [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+        ctx_id as u32
+    };
+
+    {
+        // Configure the number of vCPUs (4) and the amount of RAM (4096 MiB).
+        //
+        // SAFETY: safe as no pointers involved
+        let err = unsafe { krun_set_vm_config(ctx_id, 4, 4096) };
+        if err < 0 {
+            let err = Errno::from_raw_os_error(-err);
```
Comment (slp) [inline src/bin/krun.rs:62]: How would you feel about replacing this kind of error handling pattern with something like this?

```
            Err(anyhow!(
                "Error configuring the number of vCPUs and/or the amount of RAM: {}",
                Errno::from_raw_os_error(-err)
            ))?;
```

File: Cargo.toml (hunk 3 above)
Comment (slp) [reply Cargo.toml:15]: That would imply that packaging `krun` becomes dependent on packaging `bpaf` first. It's additional maintenance burden, but if you're up to it, that's fine by me.

**Comment by slp** [comment]
> > Is there anything else remaining to be done apart for the fork+exec replacement for passt?
> 
> There's the hardcoded `/home/{username}` for `HOME` env var. Should we pass in the actual path from `getpwnam_r` to `krun-guest` instead (already used in `krun`)?

That would be better indeed.

> There's also the TODO that you've noted about only overriding the mesa driver selection on Asahi Linux. Is it possible for us to set that in `krun` instead (through `envp` in `krun_set_exec`)?

Yes. We can probably generalize this by having an argument in `krun` that allows injecting environment variables into the guest without the need for them to be exported in the host, so the user can do something like:

`krun -e MESA_LOADER_DRIVER_OVERRIDE=asahi /bin/bash`

We can also do this later, doesn't need to be part of this PR. As you prefer.

> Other than that, it's housekeeping stuff like updating the GitHub Actions workflows and removing the old code etc. (if you choose to keep the commit history).

Ack. IMO keeping the (very short) commit history after a full rewrite that switches to another language creates confusion more than anything else. I'll create the repo either later today or tomorrow.

File: src/bin/krun.rs (hunk 4 above)
Comment (teohhanhui) [reply src/bin/krun.rs:62]: It should just be `Err(err).context("...")`

I'll change it.

**Comment by teohhanhui** [comment]
Switched to using a Cargo workspace. This will help enforce a clean separation between the `krun` and `krun-guest` binaries.

**Comment by teohhanhui** [comment]
> overriding the mesa driver selection on Asahi Linux

I was looking into https://registry.khronos.org/EGL/extensions/MESA/EGL_MESA_query_driver.txt but that seems like too much work for now...

**Comment by teohhanhui** [comment]
> `krun -e MESA_LOADER_DRIVER_OVERRIDE=asahi /bin/bash`

On that point... Do you need me to restore the hidden short options? (`-N`, `-P`)

Or were those only there because of how `getopt_long` works? (I couldn't find anything in the docs that suggests a way to disable short options, but I'm not so familiar with C, so I could be mistaken...)

**Comment by teohhanhui** [comment]
> There's the hardcoded `/home/{username}` for `HOME` env var. Should we pass in the actual path from `getpwnam_r` to `krun-guest` instead (already used in `krun`)?

Never mind. It seems like we can just do the same call from `krun-guest`. It's cleaner and allows us more flexibility in the future...

**Comment by teohhanhui** [comment]
Added `tracing` for better debuggability.

File: crates/krun/src/bin/krun.rs [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+                        .signal()
+                        .expect("either one of status code or signal should be set")
+                )
+            };
+            Err(err)?;
+        }
+
+        let is_asahi = std::str::from_utf8(&output.stdout)
+            .context("Failed to process stdout from `eglinfo` as it contains invalid UTF-8")?
+            .contains("EGL driver name: asahi");
```
Comment (teohhanhui) [inline crates/krun/src/bin/krun.rs:242]: Maybe we really shouldn't do this lol... :sweat_smile:

File: src/net.rs (hunk 1 above)
Comment (teohhanhui) [reply src/net.rs]: Done.

**Comment by teohhanhui** [comment]
There are no more remaining tasks other than setting up the CI workflow, which I intend to do next.

But perhaps we can get this into the new repo first, if you'd like.

**Comment by slp** [comment]
> There are no more remaining tasks other than setting up the CI workflow, which I intend to do next.
> 
> But perhaps we can get this into the new repo first, if you'd like.

Great! Could you please rebase this PR into the new repo?

**Comment by teohhanhui** [comment]
Done.

**Pull request closed without merging**